stdlib/asm/collections/smt.masm

```diff
@@ -301,3 +308,415 @@ export.get
+    push.EMPTY_48_3 assert_eq
+    push.EMPTY_48_2 assert_eq
+    push.EMPTY_48_1 assert_eq
+    push.EMPTY_48_0 assert_eq
+    # => [P_new, X, X, R, ...]
+
+    # prepare the stack for mtree_set operation against R; here we load idx_hi from loc[0][0]
+    # (7 cycles)
+    swapw swapw.3 loc_load.0 push.32
+    # => [32, idx_hi, R, P_new, X, X, ...]
+
+    # insert P_new into tree with root R at depth 32 and idx_hi index (29 cycles)
+    mtree_set
+    # => [P_old, R_new, X, X, ...]
+
+    # load previously saved P with P_old to make sure they are the same (6 cycles)
+    swapw swapw.3 loc_loadw.1
+    # => [P, P_old, X, R_new, ...]
+
+    # make sure P and P_old are the same (12 cycles)
+    repeat.4
+        dup.4 assert_eq
+    end
+    # => [P, X, R_new, ...]
+
+    # load values A and B from the advice provider and prepare to compute H = hash([A, B]); we'll
+    # use this to prove that P is an internal node (11 cycles)
+    swapw adv_loadw padw swapw adv_push.4
+    # => [B, A, 0, 0, 0, 0, P, R_new, ...]
+
+    # compute H = hash([A, B]) (10 cycles)
+    hperm dropw swapw dropw
+    # => [H, P, R_new, ...]
+
+    # make sure P and H are the same; this proves that the node is an internal node because leaf
+    # nodes are always hashed in a specific domain (11 cycles)
+    assert_eqw
+    # => [R_new, ...]
+
+    # put the return value onto the stack and return (4 cycles)
+    padw
+    # => [0, 0, 0, 0, R_new, ...]
+end
+
+#! Inserts the specified value into a Sparse Merkle Tree with the specified root under the
+#! specified key.
+#!
+#! The value previously stored in the SMT under this key is left on the stack together with
+#! the updated tree root.
+#!
+#! This assumes that the value is not [ZERO; 4]. If it is, the procedure fails.
+#!
+#! Input: [V, K, R, ...];
+#! Output:[V_old, R', ...]
+#!
+#! Cycles:
+#! - Update existing leaf:
+#!   - Depth 16: 129
+#!   - Depth 32: 126
+#!   - Depth 48: 131
+#! - Insert new leaf:
+#!   - Depth 16: 100
+#!   - Depth 32: 210
+#!   - Depth 48: 210
+#! - Replace a leaf with a subtree:
+#!   - Depth 32: TODO
+#!   - Depth 48: TODO
+export.insert
+    # make sure the value is not [ZERO; 4] (17 cycles)
+    repeat.4
+        dup.3 eq.0
+    end
+    and and and assertz
+    # => [V, K, R, ...]
+
+    # invoke adv and fetch target depth flags (4 cycles)
+    adv.push_smtinsert adv_push.4
+    # => [TODO]
+
+    # call the inner procedure depending on the type of insert and depth
+    if.true # --- update leaf -----------------------------------------------------
+        if.true 
+            if.true # --- update a leaf node at depth 16 ---
+                drop
+
+                # (cycles 8)
+                dup.4 exec.get_top_16_bits
+                push.16
+                # => [16, idx, V, K, R, ...]
+
+                exec.update_16_32_48
+            else # --- update a leaf node at depth 32 ---
+                drop
+
+                #(5 cycles)
+                dup.4 u32split swap drop
```

Comment:
for readability and consistency with other branches we could consider introducing a `get_top_32_bits` proc.

Reply:
Good idea!